packages/ra-core/src/controller/field/useReferenceArrayFieldController.ts

```diff
@@ -55,26 +47,235 @@ interface Option {
+            stateParams.current.sort = initialSort;
+            setSortObject(initialSort);
+        }
+    }, [initialSort, setSortObject]);
+
+    // selection logic
+    const [selectedIds, setSelectedIds] = useSafeSetState<Identifier[]>([]);
+    const onSelect = useCallback(
+        (newIds: Identifier[]) => {
+            setSelectedIds(newIds);
+        },
+        [setSelectedIds]
+    );
+    const onToggleItem = useCallback(
+        (id: Identifier) => {
+            setSelectedIds(previousState => {
+                const index = previousState.indexOf(id);
+                if (index > -1) {
+                    return [
+                        ...previousState.slice(0, index),
+                        ...previousState.slice(index + 1),
+                    ];
+                } else {
+                    return [...previousState, id];
+                }
+            });
+        },
+        [setSelectedIds]
+    );
+    const onUnselectItems = useCallback(() => {
+        setSelectedIds([]);
+    }, [setSelectedIds]);
+
+    // filter logic
+    const [displayedFilters, setDisplayedFilters] = useSafeSetState<{
+        [key: string]: boolean;
+    }>({});
+    const [filterValues, setFilterValues] = useSafeSetState<{
+        [key: string]: any;
+    }>(filter);
+    const hideFilter = useCallback(
+        (filterName: string) => {
+            setDisplayedFilters(previousState => {
+                const { [filterName]: _, ...newState } = previousState;
+                return newState;
+            });
+            setFilterValues(previousState => {
+                const { [filterName]: _, ...newState } = previousState;
+                return newState;
+            });
+        },
+        [setDisplayedFilters, setFilterValues]
+    );
+    const showFilter = useCallback(
+        (filterName: string, defaultValue: any) => {
+            setDisplayedFilters(previousState => ({
+                previousState,
+                [filterName]: true,
+            }));
+            setFilterValues(previousState => ({
+                previousState,
+                [filterName]: defaultValue,
+            }));
+        },
+        [setDisplayedFilters, setFilterValues]
+    );
+    const setFilters = useCallback(
+        (filters, displayedFilters) => {
+            setFilterValues(removeEmpty(filters));
+            setDisplayedFilters(displayedFilters);
+            setPage(1);
+        },
+        [setDisplayedFilters, setFilterValues, setPage]
+    );
+    // handle filter prop change
+    useEffect(() => {
+        if (!isEqual(filter, stateParams.current.filter)) {
+            stateParams.current.filter = filter;
+            setFilterValues(filter);
+        }
+    });
+
+    // We do all the data processing (filtering, sorting, paginating) client-side
+    useEffect(() => {
+        if (!loaded) return;
+        let finalData = data;
+        // 1. filter
+        Object.keys(filterValues).forEach(
```

Comment:
I think it would be better to start form the data, with a `data.filter` that would then check all filter values.
This way no need to duplicate data.

Additionally, you could use ramda whereEq to filter data
https://ramdajs.com/docs/#whereEq
I checked, ramda is already in react-admin dependency (well in yarn.lock actually)

Reply:
TIL ramda.wherEq

Yes, it would prevent iterating several times over the data for multiple filters. BUT Ramda isn't a direct dependency, it's significantly harder to code, and I doubt many people will download very large arrays and do multicriteria search on them. And when it's the case, we'll investigate your optimization.

Reply:
Not that hard to code:
```js
const tempData = data.filter(record =>
    Object.entries(filterValues)
        .all(([filterName, filterValue]) => filterValue == get(record, filterName)),
);
```